### PR TITLE
Save statistics for different agents

### PR DESCRIPTION
**Pull request by sdvendramini**
|Related issue|
|---|
https://github.com/wazuh/wazuh/issues/13102
||

## Description

This PR add changes for classify manager statistics into connected agents.

When a new agent connects to the manager, all of their activity will be tracked in that agent's local counters. Once the agent logs out, these statistics will be reset.

This statistics will be stored in memory and the API should be able to query them. Stat counters will reset after restarting the process.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade
- [x] Review logs syntax and correct language

- Memory tests for Linux
  - [ ] Scan-build report
  - [ ] Valgrind (memcheck and descriptor leaks check)

- [x] Added unit tests (for new features)

### CODE REVIEW

File: src/analysisd/state.h [hunk 1]
```diff
@@ -130,47 +141,47 @@ void w_inc_received_events();
 /**
  * @brief Increment syscheck decoded events counter
  */
-void w_inc_syscheck_decoded_events();
+void w_inc_syscheck_decoded_events(char * agent_id);
```
Comment (TomasTurina) [inline src/analysisd/state.h:144]: Add the `agent_id` parameter to the function description. Repeat this in all the functions.

File: src/analysisd/state.c [hunk 2]
```diff
@@ -35,6 +37,81 @@ static void w_get_queues_size();
  */
 static void w_get_initial_queues_size();
 
+/**
+ * @brief Increment syscheck decoded events counter for agents
+ */
+static void w_inc_syscheck_agent_decoded_events(char * agent_id);
```
Comment (TomasTurina) [inline src/analysisd/state.c:43]: Add the `agent_id` parameter to the function description. Repeat this in all the functions.

File: src/analysisd/state.c [hunk 3]
```diff
@@ -319,6 +397,160 @@ void w_get_initial_queues_size() {
...
+}
+
+static void w_inc_firewall_agent_written(char * agent_id) {
+    w_mutex_lock(&agents_state_mutex);
+    analysisd_agent_state_t *agent_node = get_node(agent_id);
+    agent_node->firewall_written++;
+    w_mutex_unlock(&agents_state_mutex);
+}
+
+static analysisd_agent_state_t * get_node(char *agent_id){
```
Comment (TomasTurina) [inline src/analysisd/state.c:491]:
```suggestion
static analysisd_agent_state_t * get_node(char *agent_id) {
```

File: src/analysisd/state.c [hunk 4]
```diff
@@ -319,6 +397,160 @@ void w_get_initial_queues_size() {
...
+
+            int exist = 0;
+            for (size_t i = 0; active_agents[i] != -1; i++) {
+                if (agent_state->id == active_agents[i] ) {
+                    exist = 1;
+                    break;
+                }
+            }
+
+            if (exist == 0){
```
Comment (TomasTurina) [inline src/analysisd/state.c:538]:
```suggestion
            if (exist == 0) {
```

File: src/analysisd/state.c [hunk 5]
```diff
@@ -319,6 +397,160 @@ void w_get_initial_queues_size() {
...
+    w_mutex_unlock(&agents_state_mutex);
+}
+
+static analysisd_agent_state_t * get_node(char *agent_id){
+    analysisd_agent_state_t * agent_state = (analysisd_agent_state_t *) OSHash_Get_ex(analysisd_agents_state, agent_id);
+
+    if(agent_state != NULL) {
+        return agent_state;
+    } else {
+        os_calloc(1,sizeof(analysisd_agent_state_t),agent_state);
```
Comment (TomasTurina) [inline src/analysisd/state.c:497]:
```suggestion
        os_calloc(1, sizeof(analysisd_agent_state_t), agent_state);
```

File: src/analysisd/state.c [hunk 6]
```diff
@@ -319,6 +397,160 @@ void w_get_initial_queues_size() {
...
+    }
+}
+
+static void w_analysisd_clean_agents_state() {
+    int *active_agents = NULL;
+    int sock = -1;
+    OSHashNode *hash_node;
+    unsigned int inode_it = 0;
+
+    w_mutex_lock(&agents_state_mutex);
```
Comment (TomasTurina) [inline src/analysisd/state.c:510]: Can we first get the information from the DB and at the end lock and update the counters structure?

File: src/analysisd/analysisd.c [hunk 7]
```diff
@@ -971,6 +971,15 @@ void OS_ReadMSG_analysisd(int m_queue)
 
     mdebug1("FTS_Init completed.");
 
+    /* Create OSHash for agents statistics */
+    analysisd_agents_state = OSHash_Create();
+    if (!analysisd_agents_state) {
+        merror(MEM_ERROR, errno, strerror(errno));
```
Comment (TomasTurina) [inline src/analysisd/analysisd.c:977]: I think you should use `merror_exit`

File: src/analysisd/analysisd.c [hunk 8]
```diff
@@ -971,6 +971,15 @@ void OS_ReadMSG_analysisd(int m_queue)
...
+    }
+    if (!OSHash_setSize(analysisd_agents_state, 2048)) {
+        merror(LIST_ERROR);
```
Comment (TomasTurina) [inline src/analysisd/analysisd.c:980]: Same here. Also, use a more representative message.

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/analysisd/state.c (hunk 2 above)
Comment (sdvendramini) [reply src/analysisd/state.c:44]: Done!

File: src/analysisd/state.c (hunk 4 above)
Comment (sdvendramini) [reply src/analysisd/state.c]: Done!

File: src/analysisd/state.c (hunk 5 above)
Comment (sdvendramini) [reply src/analysisd/state.c]: Done!

File: src/analysisd/analysisd.c (hunk 8 above)
Comment (sdvendramini) [reply src/analysisd/analysisd.c]: Done!

File: src/remoted/state.h [hunk 9]
```diff
@@ -13,6 +13,8 @@
 
 #include <stdint.h>
 
+OSHash *remoted_agents_state;
```
Comment (TomasTurina) [inline src/remoted/state.h:16]: Move this to .c file.

File: src/remoted/state.c [hunk 10]
```diff
@@ -13,12 +13,97 @@
...
 static pthread_mutex_t state_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_mutex_t agents_state_mutex = PTHREAD_MUTEX_INITIALIZER;
 static int rem_write_state();
 static char *refresh_time;
 
+/**
+ * @brief Search or create and return agent state node
+ * @param agent_id Id of the agent that corresponds to the node
+ */
+static remoted_agent_state_t * get_node(char *agent_id);
```
Comment (TomasTurina) [inline src/remoted/state.c:28]: Add return value to description.

File: src/remoted/state.c [hunk 11]
```diff
@@ -119,6 +205,149 @@ int rem_write_state() {
...
+
+    hash_node = OSHash_Begin(remoted_agents_state, &inode_it);
+
+    if (hash_node == NULL) {
+        return;
+    }
+
+    active_agents = wdb_get_agents_by_connection_status(AGENT_CS_ACTIVE, &sock);
+    if(!active_agents) {
+        merror("Unable to get connected agent's.");
```
Comment (TomasTurina) [inline src/remoted/state.c:234]:
```suggestion
        merror("Unable to get connected agents.");
```

File: src/remoted/state.c [hunk 12]
```diff
@@ -119,6 +205,149 @@ int rem_write_state() {
...
+                break;
+            }
+        }
+
+        if (exist == 0) {
+            agent_state = (remoted_agent_state_t *)OSHash_Delete_ex(remoted_agents_state, agent_id);
+            if (agent_state) {
+                os_free(agent_state);
+            }
+            hash_node = OSHash_Begin(remoted_agents_state, &inode_it);
```
Comment (TomasTurina) [inline src/remoted/state.c:258]: Do the same that was suggested in analysisd.

File: src/analysisd/state.c [hunk 13]
```diff
@@ -673,5 +968,81 @@ cJSON* asys_create_state_json() {
...
+            cJSON_AddNumberToObject(_statistics, "alerts_written", data->alerts_written);
+            cJSON_AddNumberToObject(_statistics, "firewall_written", data->firewall_written);
+            cJSON_AddNumberToObject(_statistics, "archives_written", data->archives_written);
+
+            cJSON_AddItemToArray(_array, _item);
+
+            hash_node = OSHash_Next(analysisd_agents_state, &index, hash_node);
+        }
+
+        cJSON_AddItemToObject(asys_state_json, "agents_conected", _array);
```
Comment (TomasTurina) [inline src/analysisd/state.c:1043]:
```suggestion
        cJSON_AddItemToObject(asys_state_json, "agents_connected", _array);
```

File: src/remoted/state.c [hunk 14]
```diff
@@ -315,5 +593,64 @@ cJSON* rem_create_state_json() {
...
+            cJSON_AddNumberToObject(_messages_sent_breakdown, "cfga_messages", data->sent_breakdown.cfga_count);
+            cJSON_AddNumberToObject(_messages_sent_breakdown, "request_messages", data->sent_breakdown.request_count);
+            cJSON_AddNumberToObject(_messages_sent_breakdown, "discarded_messages", data->sent_breakdown.discarded_count);
+
+            cJSON_AddItemToArray(_array, _item);
+
+            hash_node = OSHash_Next(remoted_agents_state, &index, hash_node);
+        }
+
+        cJSON_AddItemToObject(rem_state_json, "agents_conected", _array);
```
Comment (TomasTurina) [inline src/remoted/state.c:651]:
```suggestion
        cJSON_AddItemToObject(rem_state_json, "agents_connected", _array);
```

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/remoted/state.c (hunk 10 above)
Comment (sdvendramini) [reply src/remoted/state.c:28]: Done!

File: src/wazuh_db/wdb.h [hunk 15]
```diff
@@ -2168,7 +2169,7 @@ int wdb_global_reset_agents_connection(wdb_t *wdb, const char *sync_status);
  * @retval JSON with agents IDs on success.
  * @retval NULL on error.
  */
-cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id, const char* connection_status, wdbc_result* status);
+cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id, const char* connection_status, wdbc_result* status, char* node_name);
```
Comment (TomasTurina) [inline src/wazuh_db/wdb.h:2172]: Add new parameter to function description.

File: src/wazuh_db/helpers/wdb_global_helpers.h [hunk 16]
```diff
@@ -295,7 +296,7 @@ int wdb_reset_agents_connection(const char *sync_status, int *sock);
  * @param[in] sock The Wazuh DB socket connection. If NULL, a new connection will be created and closed locally.
  * @return Pointer to the array, on success. NULL on errors.
  */
-int* wdb_get_agents_by_connection_status(const char* connection_status, int *sock);
+int* wdb_get_agents_by_connection_status(const char* connection_status, int *sock, char* node_name);
```
Comment (TomasTurina) [inline src/wazuh_db/helpers/wdb_global_helpers.h:299]: Add new parameter to function description.

File: src/wazuh_db/wdb_parser.c [hunk 17]
```diff
@@ -5914,9 +5915,13 @@ int wdb_parse_global_get_agents_by_connection_status(wdb_t* wdb, char* input, ch
     }
     connection_status = next;
 
+    /* Get node name */
+    next = strtok_r(NULL, delim, &savedptr);
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_parser.c:5919]:
```suggestion
    node_name = strtok_r(NULL, delim, &savedptr);
```
You can simplify two lines into one.

File: src/wazuh_db/wdb.h (hunk 15 above)
Comment (TomasTurina) [reply src/wazuh_db/wdb.h:2172]: Also, I think you can make `node_name` const.

File: src/wazuh_db/helpers/wdb_global_helpers.h (hunk 16 above)
Comment (TomasTurina) [reply src/wazuh_db/helpers/wdb_global_helpers.h:299]: Also, I think you can make `node_name` const.

File: src/wazuh_db/wdb_global.c [hunk 18]
```diff
@@ -1820,19 +1820,30 @@ int wdb_global_reset_agents_connection(wdb_t *wdb, const char *sync_status) {
...
+        }
+        stmt = wdb->stmt[WDB_STMT_GLOBAL_GET_AGENTS_BY_CONNECTION_STATUS];
+    } else {
+        if (wdb_stmt_cache(wdb, WDB_STMT_GLOBAL_GET_AGENTS_BY_CONNECTION_STATUS_AND_NODE) < 0) {
+            mdebug1("Cannot cache statement");
+            *status = WDBC_ERROR;
+            return NULL;
+        }
+        stmt = wdb->stmt[WDB_STMT_GLOBAL_GET_AGENTS_BY_CONNECTION_STATUS_AND_NODE];
+        //si cambio el orden de los parametros en la query puedo meter aca el sqlite3_bind_text del node_name
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_global.c:1845]: Remember to delete comment.

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/wazuh_db/helpers/wdb_global_helpers.h (hunk 16 above)
Comment (sdvendramini) [reply src/wazuh_db/helpers/wdb_global_helpers.h]: Done!

File: src/wazuh_db/wdb.h (hunk 15 above)
Comment (sdvendramini) [reply src/wazuh_db/wdb.h]: Done!

**Pull request merged**